### DISCUSSION TITLE
Real-time input updates, and compiled model for use in HW in the loop system

### DISCUSSION THREAD

**Discussion by faisal-shah** [Hopsan/hopsan#2207] [Q&A]
I want to create a model which will be used as part of hardware in the loop testing. Ideally, I want to create a shared lib of the model which will be linked to, and call something like singleStep(), getSimState(), updateInputs().

I learned that many Simulink Simscape components do not allow for real-time updating of parameters during a simulation if you export the model as C source code. Is there any such limitation in the generated code from Hopsan? Can I modify Signal/Sources on the fly?

Are there examples or tutorials to show how to accomplish this?

Appreciate the help, thanks.

**Comment by faisal-shah** [comment]
Sooo, here's what I've learned so far. I can export an executable from the GUI, and hopsan creates a whole build directory complete with build and link scripts. Very nice and clean.

Looking at the generated exec_main.cpp, I can see the required setup/run/teardown process of running the model. It seems that I can call `hopsan::ComponentSystem::simulate` repeatedly with increasing `stopT` and get the results between calls. So, if my simulation timestep is 1ms, and I need to update inputs and get outputs every 10ms, I would run simulate() in 10ms increments and update parameters and get outputs in between.

There is a mechanism to change parameters between calls to simulate() (`setParameterValue`, and `setConstantValue`). Is this the correct way to change parameters? `setParameterValue` calls `setConstantValue` and then _reconfigures_ the component if needed and if the component requires reconfiguration. I can't find any components that actually set this to _true_. In the comments for `setConstantValue`, it says not to use this function during simulation and that it is slow. Does that mean while simulate() is running it is slow - or even between calls to simulate()? Also, I'm curious - why is it slow? It's doing a string lookup and modifying a parameter it seems - I guess you don't want to be doing that at say 10khz?

**Comment by robbr48** [comment]
This sounds like an interesting project! We have previously used Hopsan for real-time simulation mostly using S-functions exported to Simulink or by exporting Veristand DLLs to LabVIEW. I think you are on the right track, assuming you want to control the simulation directly using C++ functions. If you have some target environment available that supports the FMI standard, you might be able to export an FMU file instead.

As you mention, the `setParameterValue` function is slow because it uses name lookup and performs some error checking. If you want to change variables during simulation, it is recommended that you fetch a pointer to the variable using the `getSafeNodeDataPtr` function. You can then write the value directly to the target variable without the need for calling a function.

Reconfiguring components are not necessary. There are one or two special components that e.g. changes the number of ports depending on a parameter value, but I doubt you will be using any of those.

Also keep in mind that you can only change _input variables_, not _constants_, during simulation. Constants are only read once when starting the simulation, so changing them later will have no effect.

**Reply by faisal-shah** [reply]
Awesome, thanks! Man this project is a gem. I have a small window of opportunity to introduce hopsan in my company (as opposed to going with the default option of Simulink :) ), and hopefully be able to contribute back. I have some further questions that I'll post in this discussion forum.
